skimage/util/_flood_fill_cy.pyx

```diff
@@ -0,0 +1,143 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+"""Cython code used in _flood_fill.py."""
+
+cimport numpy as cnp
+
+
+# Must be defined to use QueueWithHistory
+ctypedef Py_ssize_t QueueItem
+
+include "../morphology/_queue_with_history.pxi"
+
+
+ctypedef fused dtype_t:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.uint32_t
+    cnp.uint64_t
+    cnp.int8_t
+    cnp.int16_t
+    cnp.int32_t
+    cnp.int64_t
+    cnp.float32_t
+    cnp.float64_t
+
+
+# Definition of flag values used for `flags` in _flood_fill & _fill_plateau
+cdef:
+    # Border value - do not cross!
+    unsigned char BORDER = 2
+    # Part of the flood fill
+    unsigned char FILL = 1
+    # Not checked yet
+    unsigned char UNKNOWN = 0
+
+
+cpdef inline void _flood_fill_equal(dtype_t[::1] image, 
+                                    unsigned char[::1] flags,
+                                    Py_ssize_t[::1] neighbor_offsets, 
+                                    Py_ssize_t start_index, 
+                                    dtype_t seed_value) nogil:
+    """Fill connected areas with 1, requiring strict equality.
+
+    Parameters
+    ----------
+    image : ndarray, one-dimensional
+        The raveled view of a n-dimensional array.
+    flags : ndarray, one-dimensional
+        An array of flags that is used to store the state of each pixel during
+        evaluation.
+    neighbor_offsets : ndarray
+        A one-dimensional array that contains the offsets to find the
+        connected neighbors for any index in `image`.
+    start_index : int
+        Start position for the flood-fill.
+    seed_value :
+        Value of `image[start_index]`.
+    """
+    cdef:
+        QueueWithHistory queue
+        QueueItem current_index, neighbor
+
+    # Initialize the queue and push start position
+    queue_init(&queue, 64)
+    queue_push(&queue, &start_index)
+    flags[start_index] = FILL
```

Comment:
I would really recommend moving the above 2 lines into the try-finally-block as well. That way we can be sure the memory is released even if something goes wrong in these lines. Applies to the other function below as well.